// @kokomi/w3g-parser-browser/dist/lib/ActionParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionParser = exports.VariableType = exports.processSaveGame = void 0;
const ByteBuffer_1 = require("./ByteBuffer");
const processPauseGame = (bb) => {
    return {
        type: 0x01,
    };
};
const processResumeGame = (bb) => {
    return {
        type: 0x02,
    };
};
const processSetGameSpeed = (bb) => {
    return {
        type: 0x03,
        speed: bb.readUint8(),
    };
};
const processIncreaseSpeed = (bb) => {
    return {
        type: 0x04,
    };
};
const processDecreaseSpeed = (bb) => {
    return {
        type: 0x05,
    };
};
const processSaveGame = (bb) => {
    return {
        type: 0x06,
        fileName: bb.readCString(),
    };
};
exports.processSaveGame = processSaveGame;
const processSaveGameFinished = (bb) => {
    return {
        type: 0x07,
        unknown: bb.readUint32(),
    };
};
const processAbilityAction = (bb) => {
    return {
        type: 0x10,
        flags: bb.readUint16(),
        itemId: bb.readUint32(),
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
    };
};
var VariableType;
(function (VariableType) {
    VariableType[VariableType["TYPE_NOTHING"] = 0] = "TYPE_NOTHING";
    VariableType[VariableType["TYPE_UNKNOWN"] = 1] = "TYPE_UNKNOWN";
    VariableType[VariableType["TYPE_NULL"] = 2] = "TYPE_NULL";
    VariableType[VariableType["TYPE_CODE"] = 3] = "TYPE_CODE";
    VariableType[VariableType["TYPE_INTEGER"] = 4] = "TYPE_INTEGER";
    VariableType[VariableType["TYPE_REAL"] = 5] = "TYPE_REAL";
    VariableType[VariableType["TYPE_STRING"] = 6] = "TYPE_STRING";
    VariableType[VariableType["TYPE_HANDLE"] = 7] = "TYPE_HANDLE";
    VariableType[VariableType["TYPE_BOOLEAN"] = 8] = "TYPE_BOOLEAN";
    VariableType[VariableType["TYPE_INTEGER_ARRAY"] = 9] = "TYPE_INTEGER_ARRAY";
    VariableType[VariableType["TYPE_REAL_ARRAY"] = 10] = "TYPE_REAL_ARRAY";
    VariableType[VariableType["TYPE_STRING_ARRAY"] = 11] = "TYPE_STRING_ARRAY";
    VariableType[VariableType["TYPE_HANDLE_ARRAY"] = 12] = "TYPE_HANDLE_ARRAY";
    VariableType[VariableType["TYPE_BOOLEAN_ARRAY"] = 13] = "TYPE_BOOLEAN_ARRAY";
})(VariableType = exports.VariableType || (exports.VariableType = {}));
const processUJApiSyncAction = (bb) => {
    bb.readUint8();
    const subtype = bb.readUint8();
    bb.readUint8();
    switch (subtype) {
        case 0x02: {
            const variableType = bb.readUint8();
            const payloadLength = bb.readUint16();
            const handleId = bb.readUint32();
            const parentKey = bb.readUint32();
            const childKey = bb.readUint32();
            let value;
            let rawValue;
            switch (variableType) {
                case 0x04:
                    value = {
                        variableType,
                        realValue: bb.readInt32(),
                    };
                    break;
                case 0x05:
                    value = {
                        variableType,
                        realValue: bb.readFloat32(),
                    };
                    break;
                case 0x06:
                    value = {
                        variableType,
                        stringValue: new TextDecoder().decode(bb.readBytes(payloadLength).toBuffer()),
                    };
                    break;
                case 0x08:
                    value = {
                        variableType,
                        booleanValue: bb.readFloat32(),
                    };
                    break;
                default:
                    rawValue = {
                        variableType,
                        data: bb.readBytes(payloadLength).toBuffer(),
                    };
            }
            return {
                type: 0xa0,
                subType: 0x02,
                payloadLength,
                handleId,
                parentKey,
                childKey,
                value,
                rawValue,
            };
        }
        default:
            bb.offset - +3;
            return undefined;
    }
};
const processPositionAbilityAction = (bb) => {
    return {
        type: 0x11,
        flags: bb.readUint16(),
        itemId: bb.readUint32(),
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
        x: bb.readUint32(),
        y: bb.readUint32(),
    };
};
const processPositionAndObjectAbilityAction = (bb) => {
    return {
        type: 0x12,
        flags: bb.readUint16(),
        itemId: bb.readUint32(),
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
        targetX: bb.readUint32(),
        targetY: bb.readUint32(),
        objectId1: bb.readUint32(),
        objectId2: bb.readUint32(),
    };
};
const processItemAction = (bb) => {
    return {
        type: 0x13,
        flags: bb.readUint16(),
        itemId: bb.readUint32(),
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
        targetX: bb.readUint32(),
        targetY: bb.readUint32(),
        objectId1: bb.readUint32(),
        objectId2: bb.readUint32(),
        itemObjectId1: bb.readUint32(),
        itemObjectId2: bb.readUint32(),
    };
};
const processAbilityTwoTargetTwoItemAction = (bb) => {
    return {
        type: 0x14,
        flags: bb.readUint16(),
        itemIdA: bb.readUint32(),
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
        targetAX: bb.readUint32(),
        targetAY: bb.readUint32(),
        itemIdB: bb.readUint32(),
        unknownC: bb.readBytes(9).toBuffer(true),
        targetBX: bb.readUint32(),
        targetBY: bb.readUint32(),
    };
};
const processChangeSelection = (bb) => {
    const mode = bb.readUint8();
    const count = bb.readUint16();
    const objects = [];
    for (let i = 0; i < count; ++i) {
        objects.push({
            objectId1: bb.readUint32(),
            objectId2: bb.readUint32(),
        });
    }
    return {
        type: 0x16,
        mode,
        objects,
    };
};
const processAssignGroupHotkey = (bb) => {
    const group = bb.readUint8();
    const count = bb.readUint16();
    const objects = [];
    for (let i = 0; i < count; ++i) {
        objects.push({
            objectId1: bb.readUint32(),
            objectId2: bb.readUint32(),
        });
    }
    return {
        type: 0x17,
        group,
        objects,
    };
};
const processSelectGroupHotkey = (bb) => {
    return {
        type: 0x18,
        group: bb.readUint8(),
        unknown: bb.readUint8(),
    };
};
const processSelectSubGroup = (bb) => {
    return {
        type: 0x19,
        itemId: bb.readUint32(),
        objectId1: bb.readUint32(),
        objectId2: bb.readUint32(),
    };
};
const processPreSubselection = (bb) => {
    return {
        type: 0x1a,
    };
};
const processUnknown1B = (bb) => {
    return {
        type: 0x1b,
        unknownA: bb.readUint8(),
        unknownB: bb.readUint32(),
        unknownC: bb.readUint32(),
    };
};
const processSelectGroudItem = (bb) => {
    return {
        type: 0x1c,
        unknownA: bb.readUint8(),
        objectId1: bb.readUint32(),
        objectId2: bb.readUint32(),
    };
};
const processCancelHeroRevival = (bb) => {
    return {
        type: 0x1d,
        unitId1: bb.readUint32(),
        unitId2: bb.readUint32(),
    };
};
const processRemoveQueuedUnit = (bb) => {
    return {
        type: 0x1e,
        slot: bb.readUint8(),
        itemId: bb.readUint32(),
    };
};
const processUnknown21 = (bb) => {
    return {
        type: 0x21,
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
    };
};
const processSinglePlayerChat = (bb) => {
    const actionId = bb.readUint8(bb.offset - 1);
    switch (actionId) {
        case 0x27:
        case 0x28:
        case 0x2d:
            return {
                type: actionId,
                unknown: bb.readUint8(),
                count: bb.readInt32(),
            };
        case 0x2e:
            return {
                type: 0x2e,
                time: bb.readFloat32(),
            };
        case 0x20:
        case 0x22:
        case 0x23:
        case 0x24:
        case 0x25:
        case 0x26:
        case 0x2a:
        case 0x2b:
        case 0x2c:
        case 0x2f:
        case 0x30:
        case 0x31:
        case 0x32:
            return {
                type: actionId,
            };
    }
    throw new Error("Unknown cheat " + actionId);
};
const processAllyOptions = (bb) => {
    return {
        type: 0x50,
        slotId: bb.readUint8(),
        flags: bb.readUint32(),
    };
};
const processResourceTransferData = (bb) => {
    return {
        type: 0x51,
        slotId: bb.readUint8(),
        gold: bb.readUint32(),
        lumber: bb.readUint32(),
    };
};
const processChatCommand = (bb) => {
    return {
        type: 0x60,
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
        command: bb.readCString(),
    };
};
const processESCKeyEvent = (bb) => {
    return {
        type: 0x61,
    };
};
const processUnknown62 = (bb) => {
    return {
        type: 0x62,
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
        unknownC: bb.readUint32(),
    };
};
const processOpenSkillSubmenu = (bb) => {
    return {
        type: 0x65,
    };
};
const processOpenBuildSubmenu = (bb) => {
    return {
        type: 0x66,
    };
};
const processMinimapPing = (bb) => {
    return {
        type: 0x68,
        x: bb.readUint32(),
        y: bb.readUint32(),
        unknown: bb.readUint32(),
    };
};
const processUnknown69 = (bb) => {
    return {
        type: 0x69,
        unknownC: bb.readUint32(),
        unknownD: bb.readUint32(),
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
    };
};
const processUnknown6A = (bb) => {
    return {
        type: 0x6a,
        unknownA: bb.readUint32(),
        unknownB: bb.readUint32(),
        unknownC: bb.readUint32(),
        unknownD: bb.readUint32(),
    };
};
const processSyncInteger = (bb) => {
    return {
        type: 0x6b,
        filename: bb.readCString(),
        missionKey: bb.readCString(),
        key: bb.readCString(),
        value: bb.readUint32(),
    };
};
const processArrowKey = (bb) => {
    return {
        type: 0x75,
        arrowKey: bb.readUint8(),
    };
};
const DEFAULT_ACTION_HANDLERS = {
    0x1: processPauseGame,
    0x2: processResumeGame,
    0x3: processSetGameSpeed,
    0x4: processIncreaseSpeed,
    0x5: processDecreaseSpeed,
    0x6: exports.processSaveGame,
    0x7: processSaveGameFinished,
    0x10: processAbilityAction,
    0x11: processPositionAbilityAction,
    0x12: processPositionAndObjectAbilityAction,
    0x13: processItemAction,
    0x14: processAbilityTwoTargetTwoItemAction,
    0x16: processChangeSelection,
    0x17: processAssignGroupHotkey,
    0x18: processSelectGroupHotkey,
    0x19: processSelectSubGroup,
    0x1a: processPreSubselection,
    0x1b: processUnknown1B,
    0x1c: processSelectGroudItem,
    0x1d: processCancelHeroRevival,
    0x1e: processRemoveQueuedUnit,
    0x20: processSinglePlayerChat,
    0x21: processUnknown21,
    0x22: processSinglePlayerChat,
    0x23: processSinglePlayerChat,
    0x24: processSinglePlayerChat,
    0x25: processSinglePlayerChat,
    0x26: processSinglePlayerChat,
    0x27: processSinglePlayerChat,
    0x28: processSinglePlayerChat,
    0x29: processSinglePlayerChat,
    0x2a: processSinglePlayerChat,
    0x2b: processSinglePlayerChat,
    0x2c: processSinglePlayerChat,
    0x2d: processSinglePlayerChat,
    0x2e: processSinglePlayerChat,
    0x30: processSinglePlayerChat,
    0x31: processSinglePlayerChat,
    0x32: processSinglePlayerChat,
    0x50: processAllyOptions,
    0x51: processResourceTransferData,
    0x60: processChatCommand,
    0x61: processESCKeyEvent,
    0x62: processUnknown62,
    0x66: processOpenSkillSubmenu,
    0x67: processOpenBuildSubmenu,
    0x68: processMinimapPing,
    0x69: processUnknown69,
    0x6a: processUnknown6A,
    0x6b: processSyncInteger,
    0x75: processArrowKey,
    0xa0: processUJApiSyncAction,
};
class ActionParser {
    constructor(actionHandler) {
        this.processActionData = (data) => {
            const bb = ByteBuffer_1.ByteBuffer.wrap(data, true);
            bb.limit = data.length;
            const actionBlocks = [];
            while (bb.remaining()) {
                const playerId = bb.readUint8();
                const actionsLength = bb.readUint16();
                const actions = [];
                const currentBlockEnd = bb.offset + actionsLength;
                while (bb.offset < currentBlockEnd) {
                    const actionId = bb.readUint8();
                    if (this.actionHandlers[actionId]) {
                        const result = this.actionHandlers[actionId](bb);
                        if (result) {
                            actions.push(result);
                        }
                        else {
                            bb.offset--;
                            break;
                        }
                    }
                    else {
                        bb.offset--;
                        break;
                    }
                }
                const remaingBuffer = bb.slice(bb.offset, currentBlockEnd).toBuffer(true);
                bb.offset = currentBlockEnd;
                actionBlocks.push({
                    playerId,
                    actions,
                    remaingBuffer,
                });
            }
            return actionBlocks;
        };
        this.actionHandlers = Object.assign(Object.assign({}, DEFAULT_ACTION_HANDLERS), actionHandler);
    }
}
exports.ActionParser = ActionParser;
